File: frontend/src/utils/uom.js
/** Base units of measure (BIZ-08 / BIZ-35). */

export const DEFAULT_UOM = 'pcs';

export const UOM_OPTIONS = [
  { value: 'pcs', label: 'Pieces (pcs)', kind: 'count' },
  { value: 'kg', label: 'Kilogram (kg)', kind: 'weight' },
  { value: 'g', label: 'Gram (g)', kind: 'weight' },
  { value: 'l', label: 'Litre (l)', kind: 'volume' },
  { value: 'ml', label: 'Millilitre (ml)', kind: 'volume' },
  { value: 'm', label: 'Metre (m)', kind: 'length' },
  { value: 'cm', label: 'Centimetre (cm)', kind: 'length' },
  { value: 'ft', label: 'Foot (ft)', kind: 'length' },
  { value: 'sqm', label: 'Square metre (sqm)', kind: 'area' },
  { value: 'sqft', label: 'Square foot (sqft)', kind: 'area' },
  { value: 'box', label: 'Box', kind: 'count' },
  { value: 'pack', label: 'Pack', kind: 'count' },
];

export function uomLabel(code) {
  return UOM_OPTIONS.find((row) => row.value === code)?.label || code || DEFAULT_UOM;
}

const MEASUREMENT_KINDS = new Set(['weight', 'volume', 'length', 'area']);

export function uomKind(code) {
  return UOM_OPTIONS.find((row) => row.value === code)?.kind || 'count';
}

export function isWeightUom(code) {
  const kind = uomKind(code);
  return kind === 'weight' || kind === 'volume';
}

export function isMeasurementUom(code) {
  return MEASUREMENT_KINDS.has(uomKind(code));
}

export function qtyStepForUom(code) {
  return isMeasurementUom(code) ? 0.001 : 1;
}

export function defaultScanQty(code) {
  return 1;
}

File: backend/app/constants/uom.py
"""Supported base units of measure (BIZ-08 / BIZ-35)."""

from __future__ import annotations

UOM_PCS = "pcs"
UOM_KG = "kg"
UOM_G = "g"
UOM_L = "l"
UOM_ML = "ml"
UOM_M = "m"
UOM_CM = "cm"
UOM_FT = "ft"
UOM_SQM = "sqm"
UOM_SQFT = "sqft"
UOM_BOX = "box"
UOM_PACK = "pack"

DEFAULT_UOM = UOM_PCS

ALLOWED_UOMS: frozenset[str] = frozenset(
    {
        UOM_PCS,
        UOM_KG,
        UOM_G,
        UOM_L,
        UOM_ML,
        UOM_M,
        UOM_CM,
        UOM_FT,
        UOM_SQM,
        UOM_SQFT,
        UOM_BOX,
        UOM_PACK,
    }
)

UOM_LABELS: dict[str, str] = {
    UOM_PCS: "Pieces",
    UOM_KG: "Kilogram",
    UOM_G: "Gram",
    UOM_L: "Litre",
    UOM_ML: "Millilitre",
    UOM_M: "Metre",
    UOM_CM: "Centimetre",
    UOM_FT: "Foot",
    UOM_SQM: "Square metre",
    UOM_SQFT: "Square foot",
    UOM_BOX: "Box",
    UOM_PACK: "Pack",
}


def normalize_uom(value: str | None, *, default: str = DEFAULT_UOM) -> str:
    if value is None or not str(value).strip():
        return default
    code = str(value).strip().lower()
    if code not in ALLOWED_UOMS:
        allowed = ", ".join(sorted(ALLOWED_UOMS))
        raise ValueError(f"Invalid unit of measure. Allowed: {allowed}")
    return code
